woocommerce/woocommerce
# Template API: Conditional visibility support

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR implements the following:
- Conditional visibility support for all blocks
- Server-side block attribute augmentation (see below for manual test that attributes exist)
- Change to how `woocommerce/conditional` block gets edited product for evaluation

Closes #40232.



### How to test the changes in this Pull Request:



#### Automated unit tests (run locally or just verify they pass in GH on this PR):

1. PHP unit tests:

```
pnpm --filter=woocommerce env:test
pnpm test:unit:env --filter 'Automattic\\WooCommerce\\Tests\\Internal\\Admin\\BlockTemplates'
```

#### Manual testing:

1. Enable the **Product Block Editor** under **WooCommerce** > **Settings** > **Advanced** > **Features** > **Experimental features**.
2. Go to **Products** > **Add New**
3. Verify each block has the `data-template-block-id` and `data-template-block-order` DOM attributes. You can do this one of two ways:
    - The hard way: view the Elements tab of the browser developer tools and look for these DOM attributes
    - The easy way: Install the [WooCommerce Block Template Inspector plugin](https://github.com/woocommerce/woocommerce-block-template-inspector) and focus (click) on each field.
4. Check that the uses of the `woocommerce/conditional` block are correct (no functional regression/change). The "Inventory" tab makes use of this block:

<details>

<summary>
Screenshots of inventory tab
</summary>

Track stock quantity disabled:

<img width="1255" alt="Screenshot 2023-10-10 at 15 10 19" src="https://github.com/woocommerce/woocommerce/assets/2098816/69bfb276-53c3-486f-ab83-be200bc61425">

Track stock quantity enabled:
<img width="1255" alt="Screenshot 2023-10-10 at 15 10 39" src="https://github.com/woocommerce/woocommerce/assets/2098816/26818e00-88bf-490a-92ef-3b6022d327b3">

</details>

5. Put the following code in an mu-plugin:

<details>

<summary>
Conditional visibility example code
</summary>

```php
use Automattic\WooCommerce\Admin\BlockTemplates\BlockInterface;

if ( ! function_exists( 'example_hook_up_block_template_conditional_visibility_modifications' ) ) {
	/**
	 * Hook up the block template after hooks.
	 */
	function example_hook_up_block_template_conditional_visibility_modifications() {
		add_action(
			'woocommerce_block_template_after_add_block',
			'example_hide_all_fields'
		);

		add_action(
			'woocommerce_block_template_area_product-form_after_add_block_product-sale-price',
			'example_hide_sale_price'
		);

		add_action(
			'woocommerce_block_template_area_product-form_after_add_block_product-fee-and-dimensions-section',
			'example_hide_shipping_fields'
		);
	}
}

add_action( 'init', 'example_hook_up_block_template_conditional_visibility_modifications', 0 );

if ( ! function_exists( 'example_hide_all_fields' ) ) {
	/**
	 * Hide the summary field.
	 *
	 * @param BlockInterface $block The block.
	 */
	function example_hide_all_fields( BlockInterface $block ) {
		if ( ! method_exists( $block, 'add_hide_condition' ) ) {
			return;
		}

		if (
			( $block->get_id() === 'product-name' ) ||
			( $block->get_id() === 'basic-details' ) ||
			( $block->get_id() === 'general' ) ) {
			// Don't hide the name field (or it's containers)!
			return;
		}

		$block->add_hide_condition( 'editedProduct.name === "boo"' );
	}
}

if ( ! function_exists( 'example_hide_sale_price' ) ) {
	/**
	 * Hide the summary field.
	 *
	 * @param BlockInterface $sale_price_block The block.
	 */
	function example_hide_sale_price( BlockInterface $sale_price_block ) {
		if ( ! method_exists( $sale_price_block, 'add_hide_condition' ) ) {
			return;
		}

		$sale_price_block->add_hide_condition(
			'editedProduct.regular_price >= 13 && editedProduct.regular_price < 14'
		);
	}
}

if ( ! function_exists( 'example_hide_shipping_fields' ) ) {
	/**
	 * Hide the summary field.
	 *
	 * @param BlockInterface $shipping_section The block.
	 */
	function example_hide_shipping_fields( BlockInterface $shipping_section ) {
		if ( ! method_exists( $shipping_section, 'add_hide_condition' ) ) {
			return;
		}

		$shipping_section->add_hide_condition(
			'editedProduct.virtual === true'
		);
	}
}
```

</details>

6. Verify the following conditional visibility happens:
    - All blocks except for the product name field are hidden when the name equals "boo"
    - The sale price field is hidden if the regular price is between $13-13.99
    - The shipping fields are hidden if the toggle at the top of the shipping tab is not enabled (marking the product as virtual)



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
